src/target/riscv/riscv.c

```diff
@@ -644,15 +653,21 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
 		return ERROR_FAIL;
 	if (riscv_reg_get(target, &tdata2_rb, GDB_REGNO_TDATA2) != ERROR_OK)
 		return ERROR_FAIL;
-	bool tdata1_config_denied = (tdata1 & ~tdata1_ignore_mask) != (tdata1_rb & ~tdata1_ignore_mask);
+	bool tdata1_config_denied = (tdata1 & ~maskmax) != (tdata1_rb & ~maskmax);
+	const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+	if (type == CSR_TDATA1_TYPE_MCONTROL &&
+	    (get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT)) {
+	    tdata1_config_denied |= (get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)))
+				    < (count_trailing_ones(tdata2) + 1));
+	}
```

Comment:
1) A third boolean flag for should be introduced here, e.g. `unsupported_napot_range`. The existing flag `tdata1_config_denied should` should not be re-used because these are two different error situations. There also should be different log messages for each (please see below).

2) Also the type of the trigger (line 657) should be taken from `tdata1_rb`. That's because the target can decide to disable the trigger (by changing the trigger type to `15`), and in that case we cannot get any of the `mcontrol` fields (`mcontrol.match`, `mcontrol.maskmax`), and all those checks should be skipped.

3) Since the conditions are quite complex, you can improve the code readability a lot by creating helper variables like `is_mcontrol`, `is_napot_match`, ...

```suggestion
	bool tdata1_config_denied = (tdata1 & ~maskmax) != (tdata1_rb & ~maskmax);
	
	bool unsupported_napot_range = false;
	const uint32_t type = get_field(tdata1_rb, CSR_TDATA1_TYPE(riscv_xlen(target)));
	const bool is_mcontrol = type == CSR_TDATA1_TYPE_MCONTROL;
	const bool is_napot_match = get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
	if (is_mcontrol && is_napot_match) {
		const riscv_reg_t maskmax = get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target));
		const unsigned napot_size = (count_trailing_ones(tdata2) + 1);
		if (maskmax < napot_size) {
			unsupported_napot_range = true;			
		}
	}
```

Error message for `unsupported_napot_range`:

```
if (tdata1_config_denied || tdata2_config_denied || unsupported_napot_range) {
	// ...

	if (unsupported_napot_range)
		LOG_TARGET_DEBUG(target,
			"The requested NAPOT match range (tdata2=0x%" PRIx64 ") exceeds maxmask=0x%" PRIx64,
			... );
	// ...
	return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
}
```

Reply:
addressed